File: OtherStuff/usecase1.js

// --- 1.
var tc = getNewTestClass();
tc.comp1.x = 7;
tc.comp2.x = 7;
tc.comp3.x = 7;

// --- 2. TestClass: js-o, TestComponent: js-o
// cpp-Code is wrong as it deletes TestComponent-instances
var tc = getNewTestClass();
var comp1 = tc.comp1; // lazy prop comp1 created

tc = null; // gc will delete tc and TestClass will delete all components
comp1.x; // crash!

// --- 2. TestClass: js-o, TestComponent: js-o
// cpp-Code is wrong as it deletes TestComponent-instances
var tc = getNewTestClass();
var comp1 = tc.comp1; // lazy prop comp1 created

comp1 = null; // gc will delete comp1
tc.comp1.x; // crash!

// --- 3. TestClass: cpp-o, TestComponent: js-o
// cpp-Code is wrong as it deletes TestComponent-instances
var tc = getNewTestClass();
var comp1 = tc.comp1; // lazy prop comp1 created

deleteTestClass(tc); // TestClass will delete all components
comp1.x; // crash!

// --- 4. TestClass: js-o, TestComponent: cpp-o
// cpp-Code is wrong as it deletes TestComponent-instances
var tc = getNewTestClass();
var comp1 = tc.comp1; // lazy prop comp1 created

tc = null; // gc will delete tc and TestClass will delete all components
comp1.x; // crash!

File: OtherStuff/usecase1.c
// lazy properties

struct TestComponent
{

	int x;
	int y;
	
	TestComponent() : x(0), y(0)
	{
	}
};

class TestClass
{
	TestComponent comp1;	// lazy prop will always point to this object
	TestComponent* comp2;	// lazy prop will point to object that was comp2 at the time
							// the prop was created
	
	public:
	TestClass()
	{
		comp2 = new TestComponent();
	}
	
	~TestClass()
	{
		delete comp2;
	}
	
	TestComponent& getComp1(){ return comp1; }
	TestComponent& getComp2(){ return comp2; }
	
	TestComponent comp3;	// lazy prop will always point to this object
};

TestClass* getNewTestClass(){ return new TestClass(); }
void deleteTestClass(TestClass* testClass){ delete testClass;}
void deleteTestComponent(TestComponent* testComp){ delete testComp;}